core/math/a_star_grid_2d.cpp

```diff
@@ -734,6 +734,90 @@ TypedArray<Vector2i> AStarGrid2D::get_id_path(const Vector2i &p_from_id, const V
 	return path;
 }
 
+Vector2 AStarGrid2D::get_flow_direction(const Vector2i &p_id) const {
+	ERR_FAIL_COND_V_MSG(dirty, Vector2(), "Grid is not initialized. Call the update method.");
+	ERR_FAIL_COND_V_MSG(!is_in_boundsv(p_id), Vector2(), vformat("Can't get point's flow direction. Point %s out of bounds %s.", p_id, region));
+	return _get_point_unchecked(p_id)->flow_direction;
+}
+
+void AStarGrid2D::set_point_cost(const Vector2i &p_id, int p_cost) {
+	ERR_FAIL_COND_MSG(dirty, "Grid is not initialized. Call the update method.");
+	ERR_FAIL_COND_MSG(!is_in_boundsv(p_id), vformat("Can't set point's cost. Point %s out of bounds %s.", p_id, region));
+	ERR_FAIL_COND_MSG(p_cost < 1 || p_cost > 254, "Can't set point's cost. Cost must be within (1-254) range.");
+	_get_point_unchecked(p_id)->cost = p_cost;
+}
+
+int AStarGrid2D::get_point_cost(const Vector2i &p_id) const {
+	ERR_FAIL_COND_V_MSG(dirty, 0, "Grid is not initialized. Call the update method.");
+	ERR_FAIL_COND_V_MSG(!is_in_boundsv(p_id), 0, vformat("Can't get point's cost. Point %s out of bounds %s.", p_id, region));
+	return _get_point_unchecked(p_id)->cost;
+}
+
+void AStarGrid2D::calculate_flow_field(const Vector2i &p_id, const Rect2i &p_region) {
+	ERR_FAIL_COND_MSG(dirty, "Grid is not initialized. Call the update method.");
+	Rect2i region_to_check;
+	if (p_region == Rect2i()) {
+		region_to_check = region;
+	} else {
+		ERR_FAIL_COND_MSG(!region.encloses(p_region), vformat("Can't calculate flow field. Region %s out of bounds %s.", p_region, region));
+		region_to_check = p_region;
+	}
+	ERR_FAIL_COND_MSG(!region_to_check.has_point(p_id), vformat("Can't calculate flow field. Point %s out of bounds %s.", p_id, region_to_check));
+
+	LocalVector<uint32_t> values;
+	values.resize(region.size.x * region.size.y);
+
+	size_t index = 0U;
+
+	// Clear previous results.
+	for (LocalVector<Point> &line : points) {
+		for (Point &p : line) {
+			p.flow_direction = Vector2();
+			values.ptr()[index++] = UINT32_MAX;
+		}
+	}
+
+	LocalVector<Point *> nbors;
+	nbors.reserve(diagonal_mode == DIAGONAL_MODE_NEVER ? 4 : 8);
+
+	LocalVector<Point *> open_list;
```

Comment:
I would probably reserve around 16 slots  to avoid the (2 - 4 - 8) reallocations.